File: codes/typescript/module/ListNode.ts
/**
 * File: ListNode.ts
 * Created Time: 2022-12-10
 * Author: Justin (xiefahit@gmail.com)
 */

/**
 * Definition for a singly-linked list node
 */
export default class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }

    /**
     * Generate a linked list with an array
     * @param arr
     * @return
     */
    arrToLinkedList(arr: number[]): ListNode | null {
        const dum: ListNode = new ListNode(0);
        let head = dum;
        for (const val of arr) {
            head.next = new ListNode(val);
            head = head.next;
        }
        return dum.next;
    }
}

File: zh-hant/codes/go/pkg/list_node.go
// File: list_node.go
// Created Time: 2022-11-25
// Author: Reanon (793584285@qq.com)

package pkg

// ListNode 鏈結串列節點
type ListNode struct {
	Next *ListNode
	Val  int
}

// NewListNode 鏈結串列節點建構子
func NewListNode(v int) *ListNode {
	return &ListNode{
		Next: nil,
		Val:  v,
	}
}

// ArrayToLinkedList 將陣列反序列化為鏈結串列
func ArrayToLinkedList(arr []int) *ListNode {
	// dummy header of linked list
	dummy := NewListNode(0)
	node := dummy
	for _, val := range arr {
		node.Next = NewListNode(val)
		node = node.Next
	}
	return dummy.Next
}
